fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/transaction/TransactionImpl.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.transfer.transaction;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
+
+import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
+
+public class TransactionImpl implements Transaction {
+	private static Thread serverThread = null;
```

Comment:
Now, maybe I'm just dumb, but why is this implemented thread-safe?
I can't think of any situation where you'd want to use transactions not on the server-thread.
While I think it's nice to have thread-safety, it only adds clutter and overhead if the rest of minecraft isn't thread-safe anyway.
If there is a good reason for all these checks and locks, please let me know. Otherwise I think this module would look a lot cleaner without it.

Reply:
You need transactions on the client thread as well to simulate transfer with items in GUIs and in the world. If someone is not careful and shares state I want to make sure that concurrent modifications are impossible.

Reply:
Use a `ThreadLocal` if you need to store the current `Transaction` instance per thread. You do NOT need to make things more safe than that. Thread safety between server and client is something modders should be used to dealing with. If modders are not "being careful" about thread safety between the client and the server then they are doing everything wrong and their mod will likely break anyway. All this locking is making everyone pay a performance cost for mistakes only a broken mod could possibly make anyway.

Reply:
Let's say you have an item in your cursor stack and you click a fluid-containing slot. Both the client and the server will try to extract fluid from the cursor stack and into the fluid-containing slot. This is not necessary for blocks as far as I know, but it will be necessary when the API is expanded to support fluid-containing items.

Reply:
I'm gonna believe you that you'd need something like this, even though I'm unsure.
But if you can do it with ThreadLocals instead (like earth suggested and I also thought of before), you should do that.
I don't see why multiple threads would have to share the same transaction stack.

Reply:
As far as I know, the entire rest of minecraft works without using any locks (except maybe in some libraries).
I don't think the risk of thread-safety is that high.
While I also don't think it's gonna be a huge performance gain to remove it, I think it would remove a lot of unnecessary clutter.

Reply:
This is really something that I would like @sfPlayer1's comment on.

Reply:
> Let's say you have an item in your cursor stack and you click a fluid-containing slot
The client and the server have *entirely different copies* of the cursor stack for *exactly this reason*. The integrated server has different references of everything like this, communication between the client and the server happens via packets, even with an integrated server.
Packets are not serialized on singleplayer, but even then, all packets not only do, but are contractually required to, create copies of all item stacks, not only for this reason, but also because in multiplayer you get thread unsafety between the network thread and the client/server thread while the network thread is serializing item stacks.

Reply:
Marking this as resolved.